File: src/remote_query.rs
use std::collections::{BTreeMap, HashSet};

use ldap3::SearchEntry;
use serde::Serialize;

use crate::{
    helpers::before_and_after_char,
    ldap_filter,
    ldap_wrapper::LdapW,
    my_err::{MyErr, Result},
    my_types::{
        CfgAndLU, Dn, RemoteConfig, RemoteDriver, RemoteSqlQuery, Subjects, ToSubjectSource,
    }, my_ldap_subjects::get_subjects,
};

#[cfg(feature = "mysql")]
mod mysql {
    use crate::{
        my_err::{MyErr, Result},
        my_types::RemoteConfig,
    };
    use mysql::prelude::Queryable;

    impl From<mysql::Error> for MyErr {
        fn from(err: mysql::Error) -> Self {
            MyErr::Msg(err.to_string())
        }
    }

    pub fn query(remote: &RemoteConfig, db_name: &str, select_query: &str) -> Result<Vec<String>> {
        let mut opts = mysql::OptsBuilder::new()
            .ip_or_hostname(Some(&remote.host))
            .db_name(Some(db_name))
            .user(Some(&remote.user))
            .pass(Some(&remote.password));
        if let Some(port) = remote.port {
            opts = opts.tcp_port(port)
        }
        let mut conn = mysql::Conn::new(opts)?;

        let rows: Vec<String> = conn.query(select_query)?;
        Ok(rows)
    }
}

#[cfg(feature = "oracle")]
mod oracle {
    use crate::my_types::RemoteConfig;
    use std::collections::Vec;

    pub fn query(remote: &RemoteConfig, select_query: &str) -> Vec<String> {
        todo!()
    }
}

fn raw_query(remote: &RemoteConfig, db_name: &str, select_query: &str) -> Result<Vec<String>> {
    match remote.driver {
        #[cfg(feature = "mysql")]
        RemoteDriver::Mysql => mysql::query(&remote, db_name, select_query),
        #[cfg(feature = "oracle")]
        RemoteDriver::Oracle => oracle::query(&remote, select_query),
    }
}

async fn sql_values_to_dns(
    ldp: &mut LdapW<'_>,
    ssdn: &Dn,
    id_attr: &str,
    sql_values: &Vec<String>,
) -> Result<HashSet<Dn>> {
    let mut r = hashset![];
    for sql_values_ in sql_values.chunks(10) {
        let filter = ldap_filter::or(
            sql_values_
                .iter()
                .map(|val| ldap_filter::eq(id_attr, val))
                .collect(),
        );
        for e in ldp.search_raw(&ssdn.0, &filter, vec![""], None).await? {
            r.insert(Dn(SearchEntry::construct(e).dn));
        }
    }
    Ok(r)
}

impl From<&ToSubjectSource> for String {
    fn from(tss: &ToSubjectSource) -> Self {
        format!("{}?{}", tss.ssdn.0, tss.id_attr)
    }
}

impl From<&RemoteSqlQuery> for String {
    fn from(rsq: &RemoteSqlQuery) -> Self {
        let opt = if let Some(subject) = rsq.to_subject_source.as_ref().map(String::from) {
            format!(" : subject={}", subject)
        } else {
            "".to_owned()
        };
        format!(
            "sql: remote={}{} : {}",
            rsq.remote_cfg_name, opt, rsq.select_query
        )
    }
}
impl From<RemoteSqlQuery> for String {
    fn from(rsq: RemoteSqlQuery) -> Self {
        String::from(&rsq)
    }
}

fn strip_prefix_and_trim<'a>(s: &'a str, prefix: &str) -> Option<&'a str> {
    s.strip_prefix(prefix).map(|s| s.trim_start())
}
fn get_param<'a>(param_name: &str, s: &'a str) -> Option<(&'a str, &'a str)> {
    eprintln!("=>>> {}", s);
    let s = s.strip_prefix(param_name)?;
    let s = s.strip_prefix('=')?;
    let (param, rest) = before_and_after_char(s, ':')?;
    Some((param.trim_end(), rest.trim_start()))
}
fn optional_param<'a>(param_name: &str, s: &'a str) -> (Option<&'a str>, &'a str) {
    if let Some((param, rest)) = get_param(param_name, s) {
        (Some(param), rest)
    } else {
        (None, s)
    }
}
fn parse_to_subject_source(s: &str) -> Result<ToSubjectSource> {
    let (ssdn, id_attr) = before_and_after_char(s, '?')
        .ok_or(MyErr::Msg(format!("expected ou=xxx,dc=xxx?uid, got {}", s)))?;
    Ok(ToSubjectSource {
        ssdn: Dn::from(ssdn),
        id_attr: id_attr.to_owned(),
    })
}
pub fn parse_sql_url(url: &str) -> Result<Option<RemoteSqlQuery>> {
    Ok(if let Some(rest) = strip_prefix_and_trim(url, "sql:") {
        let (remote_cfg_name, rest) = get_param("remote", rest)
            .ok_or(MyErr::Msg(format!("remote= is missing in {}", url)))?;
        let (subject, select) = optional_param("subject", rest);
        let to_subject_source = subject.map(parse_to_subject_source).transpose()?;
        Some(RemoteSqlQuery {
            select_query: select.to_owned(),
            remote_cfg_name: remote_cfg_name.to_owned(),
            to_subject_source,
        })
    } else {
        None
    })
}

pub fn query(
    remotes_cfg: &BTreeMap<String, RemoteConfig>,
    remote: &RemoteSqlQuery,
) -> Result<Vec<String>> {
    let remote_cfg = remotes_cfg
        .get(&remote.remote_cfg_name)
        .ok_or(MyErr::Msg(format!(
            "internal error: unknown remote {}",
            remote.remote_cfg_name
        )))?;
    raw_query(remote_cfg, &remote_cfg.db_name, &remote.select_query)
}

// return subject DNs
pub async fn query_subjects(
    ldp: &mut LdapW<'_>,
    remotes_cfg: &BTreeMap<String, RemoteConfig>,
    remote: &RemoteSqlQuery,
) -> Result<HashSet<Dn>> {
    let sql_values = query(remotes_cfg, remote)?;
    if let Some(to_ss) = &remote.to_subject_source {
        sql_values_to_dns(ldp, &to_ss.ssdn, &to_ss.id_attr, &sql_values).await
    } else {
        // the SQL query must return DNs
        Ok(sql_values.into_iter().map(Dn::from).collect())
    }
}

#[derive(Serialize)]
pub struct TestRemoteQuerySql {
    pub count: usize,
    pub values: Vec<String>,
    pub ss_guess: Option<(ToSubjectSource, Subjects)>,
}

pub async fn test_remote_query_sql(
    ldp: &mut LdapW<'_>,
    cfg_and_lu: &CfgAndLU<'_>,
    remote_sql_query: RemoteSqlQuery,
) -> Result<TestRemoteQuerySql> {
    let mut values = query(&cfg_and_lu.cfg.remotes, &remote_sql_query)?;
    let count = values.len();
    values.truncate(10); // return an extract
    let ss_guess = if count > 3 {
        guest_subject_source(ldp, &values).await?
    } else {
        None
    };
    Ok(TestRemoteQuerySql {
        count,
        values,
        ss_guess,
    })
}

async fn guest_subject_source(ldp: &mut LdapW<'_>, values: &Vec<String>) -> Result<Option<(ToSubjectSource, Subjects)>> {
    let mut best = (0, None);
    for sscfg in &ldp.config.subject_sources {
        if let Some(vec) = &sscfg.id_attrs {
            for id_attr in vec {
                let dns = sql_values_to_dns(ldp, &sscfg.dn, id_attr, values).await?;
                if dns.len() > best.0 {
                    best = (dns.len(), Some((dns, &sscfg.dn, id_attr)));
                }
            }
        }
    }
    Ok(if let Some((dns, ssdn, id_attr)) = best.1 {
        let to_subject_source = ToSubjectSource { ssdn: ssdn.clone(), id_attr: id_attr.to_owned() };
        let subjects = get_subjects(ldp, dns.into_iter().collect(), &None, &None).await?;
        Some((to_subject_source, subjects))
    } else { None })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_sql_url() -> Result<()> {
        fn test_ok(url: &str) -> Result<()> {
            let remote = parse_sql_url(url)?;
            assert_eq!(remote.map(String::from), Some(String::from(url)));
            Ok(())
        }
        fn test_err(url: &str) {
            if let Ok(remote) = parse_sql_url(url) {
                assert!(false, "unexpected success {:?}", remote);
            }
        }
        test_ok(
            "sql: remote=foo : subject=ou=people,dc=nodomain?uid : select username from users",
        )?;
        test_ok("sql: remote=foo : select concat('uid=', username, ',ou=people,dc=nodomain') from users")?;
        test_err("sql: select username from users");
        Ok(())
    }
}

File: server/remote_query.ts
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/unbound-method */
import _ from "lodash";
import mysql from 'mysql'
// @ts-expect-error (@types/oracledb 5.2.x does not allow oracledb.getConnection)
import oracledb from 'oracledb'
import { promisify } from "util";

import * as ldp from "./ldap_read_search"
import * as ldpSubject from './ldap_subject'
import conf from "./conf";
import ldap_filter from "./ldap_filter";
import { before_and_after, strip_prefix, throw_ } from "./helpers";
import { Dn, DnsOpts, hMyMap, MyMap, Option, RemoteSqlConfig, remoteSqlDrivers, RemoteSqlQuery, Subjects, toDn, ToSubjectSource } from "./my_types";
import { exact_dn_to_subject_source_cfg } from "./dn";

const driver_query: Record<string, (remote: RemoteSqlConfig, db_name: string, select_query: string) => Promise<string[]>> = {
    mysql: async (remote, db_name, select_query) => {
        const connection = mysql.createConnection({
            host     : remote.host,
            user     : remote.user,
            password : remote.password,
            port     : remote.port,
            database : db_name,
        });
        const connect = promisify(connection.connect).bind(connection)
        const query = promisify(connection.query).bind(connection)
        const end = promisify(connection.end).bind(connection)

        await connect()
        const rows = await query(select_query) as MyMap<string, unknown>[]
        await end()

        return rows.map(hMyMap.firstValue) as string[]
    },
    oracle: async (remote, db_name, select_query) => {
        const port_string = remote.port ? `:${remote.port}` : ""
        const conn = await oracledb.getConnection({ user: remote.user, password: remote.password, connectString: `${remote.host}${port_string}/${db_name}` })
        const r = await conn.execute(select_query) as { rows: string[][] }
        return r.rows.map(e => e[0])
    },
};

async function raw_query(remote: RemoteSqlConfig, db_name: string, select_query: string): Promise<string[]> {
    const f = driver_query[remote.driver]
    return await f(remote, db_name, select_query)
}

async function sql_values_to_dns_(ssdn: Dn, id_attrs: string[], sql_values: string[]) {
    const r: DnsOpts = {};
    for (const sql_values_ of _.chunk(sql_values, 10)) {
        const filter = ldap_filter.or(
            sql_values_.flatMap(val => 
                id_attrs.map(id_attr => ldap_filter.eq(id_attr, val))
            )
        );
        for (const e of await ldp.searchRaw(ssdn, filter, [""], {})) {
            r[toDn(e.dn)] = {}
        }
    }
    return r
}

const hToSubjectSource = {
    /** NB: syntax inspired by "o=Example?uid" in Apache HTTPD AuthLDAPURL */
    toString: (tss: ToSubjectSource) => (
        `${tss.ssdn}?${tss.id_attr || '*'}`
    ),
}

export const to_sql_url = (rsq: RemoteSqlQuery) => {
        const opt = rsq.to_subject_source ?
            ` : subject=${hToSubjectSource.toString(rsq.to_subject_source)}` :
            "";
        return `sql: remote=${rsq.remote_cfg_name}${opt} : ${rsq.select_query}`
}

function strip_prefix_and_trim(s: string, prefix: string): Option<string> {
    return strip_prefix(s, prefix)?.trimStart()
}
function get_param(param_name: string, s: string): Option<[string, string]> {
    console.log("=>>>", s)
    return strip_prefix(s, param_name + '=')
        ?.oMap(s => before_and_after(s, ':'))
        ?.oMap(([param, rest]) => [ param.trimEnd(), rest.trimStart() ])
}

const optional_param = (param_name: string, s: string): [Option<string>, string] => (
    get_param(param_name, s) ?? [undefined, s]
);

/**
 * @param s - string format inspired by "o=Example?uid" in Apache HTTPD AuthLDAPURL
 */
function parse_to_subject_source(s: string): ToSubjectSource {
    const [ssdn, id_attr] = before_and_after(s, '?') ?? throw_("expected ou=xxx,dc=xxx?uid, got " + s)
    return { ssdn: toDn(ssdn), id_attr: id_attr === '*' ? undefined : id_attr }
}

export const parse_sql_url = (url: string): Option<RemoteSqlQuery> => (
    strip_prefix_and_trim(url, "sql:")?.oMap(rest => {
        const [remote_cfg_name, rest_] = get_param("remote", rest) ?? throw_("remote= is missing in " + url)
        const [subject, select_query] = optional_param("subject", rest_)
        return {
            select_query,
            remote_cfg_name,
            to_subject_source: subject?.oMap(parse_to_subject_source),
        }
    })
)

const to_RemoteSqlConfig = (remote_cfg_name: string): Option<RemoteSqlConfig> => {
    const remote_cfg = conf.remotes[remote_cfg_name] ?? throw_("internal error: unknown remote " + remote_cfg_name)
    if (remoteSqlDrivers.includes(remote_cfg.driver as any)) {
        return remote_cfg as RemoteSqlConfig
    }
    return undefined
}

export function sql_query(remote: RemoteSqlQuery) {
    const remote_cfg = to_RemoteSqlConfig(remote.remote_cfg_name) ?? throw_("internal error: remote is not SQL " + remote.remote_cfg_name)
    return raw_query(remote_cfg, remote_cfg.db_name, remote.select_query)
}

const to_ss_to_id_attrs = (to_ss: ToSubjectSource) => {
    const sscfg = exact_dn_to_subject_source_cfg(to_ss.ssdn) || throw_(`invalid remote query: no id_attr and ${to_ss.ssdn} is not listed in conf.ldap.subject_sources`)
    return sscfg.id_attrs || throw_("no id_attrs for " + sscfg.name + " but needed for remote query on DN " + to_ss.ssdn)
}

export async function sql_values_to_dns(to_ss: Option<ToSubjectSource>, sql_values: string[]): Promise<DnsOpts> {
    if (to_ss) {
        const id_attrs = to_ss.id_attr ? [to_ss.id_attr] : to_ss_to_id_attrs(to_ss)
        return await sql_values_to_dns_(to_ss.ssdn, id_attrs, sql_values)
    } else {
        // the SQL query must return DNs
        return _.fromPairs(sql_values.map(dn => [dn, {}]))
    }
}

export async function guess_subject_source(values: string[]) {
    if (values.every(value => value.endsWith(conf.ldap.base_dn))) {
        // all valeurs are DNs, no subject_source needed
        return undefined
    }
    let best: Option<[number, [DnsOpts, Dn, string]]>;
    for (const sscfg of conf.ldap.subject_sources) {
        for (const id_attr of sscfg.id_attrs ?? []) {
            const dns = await sql_values_to_dns_(toDn(sscfg.dn), [id_attr], values)
            const nb_dns = _.size(dns)
            if (nb_dns > (best?.[0] ?? 0)) {
                best = [nb_dns, [dns, toDn(sscfg.dn), id_attr]];
            }
        }
    }
    if (!best) return undefined
    const [dns, ssdn, id_attr] = best[1]
    const subjects = await ldpSubject.get_subjects_(dns)
    const r: [ToSubjectSource, Subjects] = [ { ssdn, id_attr }, subjects ]
    return r
}
